proto/sql_api.proto

```diff
@@ -0,0 +1,154 @@
+  //
+  // TODO(spencer): current unimplemented.
+  CONSENSUS = 1;
+  // INCONSISTENT reads return the latest available, committed values.
+  // They are more efficient, but may read stale values as pending
+  // intents are ignored.
+  INCONSISTENT = 2;
+}
+
+// CmdRequestHeader is supplied with every CmdRequest.
+message CmdRequestHeader {
+  // Timestamp specifies time at which read or writes should be
+  // performed. If the timestamp is set to zero value, its value
+  // is initialized to the wall time of the receiving gateway node.
+  optional Timestamp timestamp = 3 [(gogoproto.nullable) = false];
+  // CmdID is optionally specified for request idempotence
+  // (i.e. replay protection).
+  optional ClientCmdID cmd_id = 4 [(gogoproto.nullable) = false, (gogoproto.customname) = "CmdID"];
+  // User is the originating user. Used to lookup priority when
+  // scheduling queued operations at target node.
+  optional string user = 5 [(gogoproto.nullable) = false];
+  // UserPriority specifies priority multiple for non-transactional
+  // commands. This value should be a positive integer [1, 2^31-1).
+  // It's properly viewed as a multiple for how likely this
+  // transaction will be to prevail if a write conflict occurs.
+  // Commands with UserPriority=100 will be 100x less likely to be
+  // aborted as conflicting transactions or non-transactional commands
+  // with UserPriority=1. This value is ignored if Txn is
+  // specified. If neither this value nor Txn is specified, the value
+  // defaults to 1.
+  optional int32 user_priority = 6 [default = 1];
+  // Txn is set non-nil if a transaction is underway. To start a txn,
+  // the first request should set this field to non-nil with name and
+  // isolation level set as desired. The response will contain the
+  // fully-initialized transaction with txn ID, priority, initial
+  // timestamp, and maximum timestamp. The transaction returned in a
+  // response is feed to the next request.
+  optional Transaction txn = 7;
+  // ReadConsistency specifies the consistency for read
+  // operations. The default is CONSISTENT. This value is ignored for
+  // write operations.
+  optional ReadConsistencyType read_consistency = 8 [(gogoproto.nullable) = false];
+}
+
+// CmdResponseHeader is returned with every Cmd response.
+message CmdResponseHeader {
+  // Error is non-nil if an error occurred.
+  optional Error error = 1;
+  // Transaction is non-nil if the request specified a non-nil
+  // transaction. The transaction timestamp and/or priority may have
+  // been updated, depending on the outcome of the request.
+  optional Transaction txn = 3;
+}
+
+// A sql CmdRequest to cockroach. A transaction can consists of multiple
+// CmdRequest.
+message CmdRequest {
+  optional CmdRequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional string sql = 2 [(gogoproto.nullable) = false];
+  // parameters are referred to in the above sql query using $1, $2, etc.
```

Comment:
The standard placeholder in SQL statements is `?`. Not sure how standard this is across DBs, but it certainly applies to MySQL. See https://dev.mysql.com/doc/refman/5.0/en/sql-syntax-prepared-statements.html.